internal/transform/contract_data.go

```diff
@@ -0,0 +1,503 @@
+
+	contractDataExpirationLedgerSeq := contractData.ExpirationLedgerSeq
+
+	transformedPool := ContractDataOutput{
+		ContractId:                  contractDataContractId.HexString(),
+		ContractKeyType:             contractDataKeyType,
+		ContractDurability:          contractDataDurability,
+		ContractDataFlags:           uint32(contractDataDataFlags),
+		ContractExpirationLedgerSeq: uint32(contractDataExpirationLedgerSeq),
+		ContractDataAssetCode:       contractDataAssetCode,
+		ContractDataAssetIssuer:     contractDataAssetIssuer,
+		ContractDataBalanceHolder:   base64.StdEncoding.EncodeToString(contractDataBalanceHolder[:]),
+		ContractDataBalance:         contractDataBalance.String(),
+		LastModifiedLedger:          uint32(ledgerEntry.LastModifiedLedgerSeq),
+		LedgerEntryChange:           uint32(changeType),
+		Deleted:                     outputDeleted,
+	}
+	return transformedPool, nil
+}
+
+// AssetFromContractData takes a ledger entry and verifies if the ledger entry
+// corresponds to the asset info entry written to contract storage by the Stellar
+// Asset Contract upon initialization.
+//
+// Note that AssetFromContractData will ignore forged asset info entries by
+// deriving the Stellar Asset Contract ID from the asset info entry and comparing
+// it to the contract ID found in the ledger entry.
+//
+// If the given ledger entry is a verified asset info entry,
+// AssetFromContractData will return the corresponding Stellar asset. Otherwise,
+// it returns empty strings.
+//
+// References:
+// https://github.com/stellar/rs-soroban-env/blob/v0.0.16/soroban-env-host/src/native_contract/token/public_types.rs#L21
+// https://github.com/stellar/rs-soroban-env/blob/v0.0.16/soroban-env-host/src/native_contract/token/asset_info.rs#L6
+// https://github.com/stellar/rs-soroban-env/blob/v0.0.16/soroban-env-host/src/native_contract/token/contract.rs#L115
+//
+// The asset info in `ContractData` entry takes the following form:
+//
+//   - Instance storage - it's part of contract instance data storage
+//
+//   - Key: a vector with one element, which is the symbol "AssetInfo"
+//
+//     ScVal{ Vec: ScVec({ ScVal{ Sym: ScSymbol("AssetInfo") }})}
+//
+//   - Value: a map with two key-value pairs: code and issuer
+//
+//     ScVal{ Map: ScMap(
+//     { ScVal{ Sym: ScSymbol("asset_code") } -> ScVal{ Str: ScString(...) } },
+//     { ScVal{ Sym: ScSymbol("issuer") } -> ScVal{ Bytes: ScBytes(...) } }
+//     )}
+func AssetFromContractData(ledgerEntry xdr.LedgerEntry, passphrase string) (string, string) {
+	contractData, ok := ledgerEntry.Data.GetContractData()
+	if !ok {
+		return "", ""
+	}
+	if contractData.Key.Type != xdr.ScValTypeScvLedgerKeyContractInstance ||
+		contractData.Body.BodyType != xdr.ContractEntryBodyTypeDataEntry {
+		return "", ""
+	}
+	contractInstanceData, ok := contractData.Body.Data.Val.GetInstance()
+	if !ok || contractInstanceData.Storage == nil {
+		return "", ""
+	}
+
+	// we don't support asset stats for lumens
```

Comment:
What are the implications of Hubble not storing asset stats for lumens? Is that ok?

Reply:
Good catch. We do want to capture the native asset. Adjusted the check to get the asset and the balance from contract data